File: algorithms/levenshtein-distance/Levenshtein.vb
Imports System

Module Levenshtein

    Function Levenshtein(ByVal w1 As String, ByVal w2 As String) As UInteger
        w1 = w1.ToLower
        w2 = w2.ToLower
        Dim n1 As UInteger = w1.Length + 1
        Dim n2 As UInteger = w2.Length + 1
        Dim d(n1, n2) As UInteger

        For y As UInteger = 0 To n1 - 1
            For x As UInteger = 0 To n2 - 1
                d(y, x) = 0
            Next
        Next
        For y As UInteger = 1 To n1 - 1
            d(y, 0) = y
        Next
        For x As UInteger = 1 To n1 - 1
            d(0, x) = x
        Next

        For y As UInteger = 1 To n1 - 1
            For x As UInteger = 1 To n2 - 1
                If w1.Chars(y - 1) = w2.Chars(x - 1) Then
                    d(y, x) = d(y - 1, x - 1)
                Else
                    d(y, x) = Math.Min(d(y - 1, x - 1), Math.Min(d(y - 1, x), d(y, x - 1))) + 1
                End If
            Next
        Next

        Levenshtein = d(n1 - 1, n2 - 1)
    End Function


    Sub Main(args As String())
        Console.WriteLine(Levenshtein("Roman", "Ramo"))
    End Sub
End Module

File: algorithms/levenshtein-distance/levenshtein.pas
program levenshtein;
uses math;

function levenshtein(w1: string; w2: string): integer;
var
	d : array of array of integer;
	n1, n2, x, y: integer;
begin
	w1 := lowercase(w1);
	w2 := lowercase(w2);
	n1 := length(w1) + 1;
	n2 := length(w2) + 1;
	
	setlength(d, n1);
	for y := 1 to n1 do
	begin
		setlength(d[y], n2);
		for x := 1 to n2 do
		begin
			d[y][x] := 0;
		end;
	end;
	for y := 2 to n1 do
	begin
		d[y][1] := y;
	end;	
	for x := 2 to n2 do
	begin
		d[1][x] := x;
	end;

	for y := 2 to n1 do
	begin
		for x := 2 to n2 do
		begin
			if w1[y-1] = w2[x-1] then d[y][x] := d[y-1][x-1]
			else d[y][x] := min(d[y-1][x-1], min(d[y-1][x], d[y][x-1])) + 1;
		end;
	end;

	levenshtein := d[n1][n2];
end;

begin
	writeln(levenshtein('Roman', 'Ramo'));
end.